### PR TITLE
#236  watch: Notify watches of connection loss

### PR DESCRIPTION
**Pull request by Grisu118**
Always send an EtcdException with Unavailable Status to active watchers
on connection loss. This fix guards against concurrency issues when creating
watches while the connection to etcd is being restored.

### CODE REVIEW

**Comment by fanminshi** [comment]
`Always send an EtcdException with Unavailable Status to active watchers
on connection loss.` I don't think the etcd go client has this behavior. that's why jetcd client doesn't throw EtcdException. But i need to double check on that.

**Comment by fanminshi** [comment]
@Grisu118 have you read the go client watch code? the jetcd watch implementation matches of that.

File: jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java [hunk 1]
```diff
@@ -94,54 +97,50 @@ public Watcher watch(ByteSequence key) {
...
 
     WatcherImpl watcher = new WatcherImpl(key, watchOption, this);
     this.pendingWatchers.add(watcher);
-
-    if (this.pendingWatchers.size() == 1) {
-      // head of the queue send watchCreate request.
-      WatchRequest request = this.toWatchCreateRequest(watcher);
-      this.getGrpcWatchStreamObserver().onNext(request);
-    }
+    this.triggerWatcherCreation();
```
Comment (fanminshi) [inline jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: what's the issue with previous implementation on 

```java
if (this.pendingWatchers.size() == 1) {		
 -      // head of the queue send watchCreate request.		
 -      WatchRequest request = this.toWatchCreateRequest(watcher);		
 -      this.getGrpcWatchStreamObserver().onNext(request);		
 -    }
```

**Comment by Grisu118** [comment]
Sorry for the late answer. I saw issue #236 and the comment of @lburgazzoli. There he wrotes about returning from listen and resubscribe on next call of listen. As this behavior was helpful for our needs I tried to implement a solution. This pull request is the second attempt to do this, as my version had a bug. Need to do some internal testing to see if we can live without that behavior.

File: jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java (hunk 1 above)
Comment (Grisu118) [reply jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java]: the issue was that with the retry logic and multiple open watchers we got concurrency issues.

**Pull request closed without merging**